exercises/practice/state-of-tic-tac-toe/test/state_of_tic_tac_toe_test.exs

```diff
@@ -0,0 +1,300 @@
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :win}
+    end
+
+    @tag :pending
+    test "Finished game where X won via diagonal victory (2)" do
+      board = """
+      O.X
+      OX.
+      X..
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :win}
+    end
+
+    @tag :pending
+    test "Finished game where O won via diagonal victory (3)" do
+      board = """
+      OXX
+      OOX
+      X.O
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :win}
+    end
+
+    @tag :pending
+    test "Finished game where O won via diagonal victory (1)" do
+      board = """
+      ..O
+      .OX
+      OXX
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :win}
+    end
+
+    @tag :pending
+    test "Finished game where X won via a row and a column victory" do
+      board = """
+      XXX
+      XOO
+      XOO
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :win}
+    end
+
+    @tag :pending
+    test "Finished game where X won via two diagonal victories" do
+      board = """
+      XOX
+      OXO
+      XOX
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :win}
+    end
+  end
+
+  describe "Drawn games" do
+    @tag :pending
+    test "Draw (1)" do
+      board = """
+      XOX
+      XXO
+      OXO
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :draw}
+    end
+
+    @tag :pending
+    test "Draw (2)" do
+      board = """
+      XXO
+      OXX
+      XOO
+      """
+
+      {:ok, :draw}
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :draw}
+    end
+  end
+
+  describe "Ongoing games" do
+    @tag :pending
+    test "Ongoing game (1)" do
+      board = """
+      ...
+      X..
+      ...
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :ongoing}
+    end
+
+    @tag :pending
+    test "Ongoing game (2)" do
+      board = """
+      O..
+      .X.
+      ...
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :ongoing}
+    end
+
+    @tag :pending
+    test "Ongoing game (3)" do
+      board = """
+      X..
+      .XO
+      OX.
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:ok, :ongoing}
+    end
+  end
+
+  describe "Invalid boards" do
+    @tag :pending
+    test "Invalid board (1)" do
+      board = """
+      XX.
+      ...
+      ...
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:error, "Wrong turn order: X went twice"}
+    end
+
+    @tag :pending
+    test "Invalid board (2)" do
+      board = """
+      OOX
+      ...
+      ...
+      """
+
+      assert StateOfTicTacToe.gamestate(board) == {:error, "Wrong turn order: O started"}
+    end
+
+    @tag :pending
+    test "Invalid board (3)" do
+      board = """
+      XXX
+      OOO
+      ...
+      """
+
+      assert StateOfTicTacToe.gamestate(board) ==
+               {:error, "Impossible board: game should have ended after X won"}
+    end
```

Comment:
I am not a fan of this test. It forces you to handle a case of two winners, but more detailed tests cases are lacking because it's impossible to handle two winners correctly in all cases. It felt wrong writing code handling this... I even left a comment in my solution. Basically in this case, assuming the right round order was followed, it's clear that X won because X started. But in this case, it's no longer clear:

```
XXX
OOO
X..
```

The real winner depends on which X was played in which order. If the bottom row X was last, then X won. But it wasn't last, then O won.

I wish this corner case of "two winners" was not even touched in the problem specs but I'm not sure if it makes sense to argue. What do you think?

Reply:
[Open a PR here](https://github.com/exercism/problem-specifications/pull/2012).

Reply:
> I was really enjoying this one until the last test (not your fault) 😞

Sorry! 😅

@jiegillet I like that change, it mirrors how I expected people to solve it (counting the number of wins, not caring about the order)